src/main/kotlin/io/github/gradlenexus/publishplugin/GetStagingProfile.kt

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright 2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.github.gradlenexus.publishplugin
+
+import io.github.gradlenexus.publishplugin.internal.NexusClient
+import org.gradle.api.GradleException
+import org.gradle.api.file.RegularFileProperty
+import org.gradle.api.model.ObjectFactory
+import org.gradle.api.tasks.Input
+import org.gradle.api.tasks.Optional
+import org.gradle.api.tasks.OutputFile
+import org.gradle.api.tasks.TaskAction
+import org.gradle.kotlin.dsl.property
+import javax.inject.Inject
+
+/**
+ * Gradle task to retrieve the staging profile id for a given [repository] and [packageGroup]
+ */
+@Suppress("UnstableApiUsage")
+open class GetStagingProfile @Inject constructor(
+    objects: ObjectFactory,
+    extension: NexusPublishExtension,
+    repository: NexusRepository
+) : AbstractNexusStagingRepositoryTask(objects, extension, repository) {
+
+    /**
+     * This property is [Optional] because it is only required if [NexusRepository.stagingProfileId] is not provided
+     */
+    @Input
+    @Optional
+    val packageGroup = objects.property<String>().apply {
+        set(extension.packageGroup)
+    }
+
+    @OutputFile
+    val stagingProfileIdOutput: RegularFileProperty = objects.fileProperty()
+
+    @TaskAction
+    fun getStagingProfile() {
+        val repository = repository.get()
+
+        val configuredStagingProfileId = repository.stagingProfileId.orNull
+        val stagingProfileId: String = if (configuredStagingProfileId == null) {
+            val client = NexusClient(
+                repository.nexusUrl.get(),
+                repository.username.orNull,
+                repository.password.orNull,
+                clientTimeout.orNull,
+                connectTimeout.orNull
+            )
+            val packageGroup = packageGroup.get()
+            client.findStagingProfileId(packageGroup) ?: throw GradleException("Failed to find staging profile for package group: $packageGroup")
+        } else {
+            configuredStagingProfileId
+        }
+
+        logger.lifecycle("Received staging profile id: '{}' for package org.example", stagingProfileId)
```

Comment:
I like the idea of reusing the `getStagingProfile` task by the initialization task. However, I see a side effect which disturbs the silence of a regular Gradle tasks executions. On every publishing there will be a few lines displayed (task name, empty line and the message). Especially, having `stagingProfileId` explicitly configured it would be unexpected.

Possibly, we could detect if the task was explicitly called from the command line (e.g. using `gradle.startParameter.taskNames`) and only then display that message at the lifecycle level (otherwise on info). That might be not very elegant, but maybe @marcphilipp knows some Gradle trick to tweak it.

In addition, `Received staging profile id` is not fully true for the explicitly configured value (it was not received) and might be confusing for people migrating from GNSP.  Maybe there could be two variants of the output:
> Configured staging profile id: ...

and 
> Received staging profile id: ...

used depending on configuration. Nevertheless, I'm not fixed on that.

Reply:
This makes sense, I agree that we don't want this to be printed out every time this task is executed. My first thought was to add a configuration property in the `GetStagingProfile` which determines whether or not the diagnostic data is printed out. That task property could be configured in the `NexusPublishPlugin`. However, I don't see a way of doing that without creating a new registration of the task with a different name to allow for the two different configurations to exist

In my opinion, that means the `GetStagingProfile` task is doing too much, as it trying to be a diagnostic tool and a first-class step in the publishing task chain. I think it would make sense to register a  `diagnosticGet${repository.name.capitalize()}StagingProfile` task for each repository that depends on the `get${repository.name.capitalize()}StagingProfile` task. The `GetStagingProfile` class is solely responsible for determining the staging profile id, either from configuration or from the HTTP API of the configured nexus repository. the `DiagnosticGetStagingProfile` consumes the staging profile ID and is instead focused on printing out the right message, depending on whether or not a `NexusRepository.stagingProfileId` is configured. This means that the `initialize` task can rely strictly upon the `GetStagingProfile` task but users still have access to a diagnostic message via the CLI if they deliberately execute the task.  This might look like:

```kotlin
/**
 * Diagnostic task for printing out the provided [stagingProfileId] in an informative message. The message
 * format differs based on whether or not the [NexusRepository.stagingProfileId] was configured explicitly
 * or not. If it was not, the [stagingProfileId] is presumed to have been obtained from a remote source,
 * like the [NexusRepository] HTTP API.
 */
@Suppress("UnstableApiUsage")
@Incubating
open class DiagnosticGetStagingProfile @Inject constructor(
    objects: ObjectFactory,
    extension: NexusPublishExtension,
    repository: NexusRepository
) : AbstractNexusStagingRepositoryTask(objects, extension, repository) {


    /**
     * This property is [Optional] because it is only required if [NexusRepository.stagingProfileId] is not provided
     */
    @Input
    @Optional
    val packageGroup = objects.property<String>().apply {
        set(extension.packageGroup)
    }

    @Input
    val stagingProfileId = project.objects.property(String::class)

    @TaskAction
    fun diagnosticGetStagingProfile() {
        val repository = repository.get()
        val configuredStagingProfileId = repository.stagingProfileId.orNull
        if (configuredStagingProfileId == null) {
            logger.lifecycle("Received staging profile id: '{}' for package '{}' for repository '{}'", stagingProfileId.get(), packageGroup.get(), repository.capitalizedName)
        } else {
            logger.lifecycle("Configured staging profile id: '{}' for repository '{}'", configuredStagingProfileId, repository.capitalizedName)
        }
    }
}
```

In `NexusPublishPlugin`, we would register the task and configure the properties and task dependencies for each repository

```kotlin
val diagnosticGetStagingProfile = rootProject.tasks.register<DiagnosticGetStagingProfile>("diagnosticGet${repository.name.capitalize()}StagingProfile")

getStagingProfileTask {
    description = "Gets a staging profile id for a given repository. If the staging profile id is configured, this task simply retrieves it from the extension that configured it. If the it is not, the staging profile id is retrieved from the NexusRepository via its HTTP API a packageGroup must be configured on the project that applied this plugin."
      group = PublishingPlugin.PUBLISH_TASK_GROUP
      stagingProfileIdOutput.set(project.layout.buildDirectory.file("publish-plugin/get${capitalizedName}StagingProfileOutput.txt"))
}
diagnosticGetStagingProfile {
    description = "Displays staging profile id for a given repository. This is a diagnostic task to get the value and put it into the NexusRepository configuration closure as stagingProfileId."
    group = PublishingPlugin.PUBLISH_TASK_GROUP
    stagingProfileId.set(getStagingProfileTask.map { it.stagingProfileIdOutput.get().asFile.readText() })
}

```

Reply:
I think we should at most log this on `info` level.

Reply:
With my [last comment](https://github.com/gradle-nexus/publish-plugin/pull/62#discussion_r579826757) the whole case with different logging would disappear.